plugins/incremental-ingestion-backend/src/engine/IncrementalIngestionEngine.ts

```diff
@@ -0,0 +1,309 @@
+        attempts: record.attempts as number,
+        nextActionAt: record.next_action_at.valueOf() as number,
+      };
+    }
+    const result = await this.manager.createProviderIngestionRecord(
+      providerName,
+    );
+    if (result) {
+      this.options.logger.info(
+        `incremental-engine: Ingestion record created: '${result.ingestionId}'`,
+      );
+    }
+    return result;
+  }
+
+  async ingestOneBurst(id: string, signal: AbortSignal) {
+    const lastMark = await this.manager.getLastMark(id);
+
+    const cursor = lastMark ? lastMark.cursor : undefined;
+    let sequence = lastMark ? lastMark.sequence + 1 : 0;
+
+    const start = performance.now();
+    let count = 0;
+    let done = false;
+    this.options.logger.info(
+      `incremental-engine: Ingestion '${id}' burst initiated`,
+    );
+
+    await this.options.provider.around(async (context: unknown) => {
+      let next = await this.options.provider.next(context, cursor);
+      count++;
+      for (;;) {
+        done = next.done;
+        await this.mark({
+          id,
+          sequence,
+          entities: next?.entities,
+          done: next.done,
+          cursor: next?.cursor,
+        });
+        if (signal.aborted || next.done) {
+          break;
+        } else {
+          next = await this.options.provider.next(context, next.cursor);
+          count++;
+          sequence++;
+        }
+      }
+    });
+
+    this.options.logger.info(
+      `incremental-engine: Ingestion '${id}' burst complete. (${count} batches in ${Math.round(
+        performance.now() - start,
+      )}ms).`,
+    );
+    return done;
+  }
+
+  async mark(options: {
+    id: string;
+    sequence: number;
+    entities?: DeferredEntity[];
+    done: boolean;
+    cursor?: unknown;
+  }) {
+    const { id, sequence, entities, done, cursor } = options;
+    this.options.logger.debug(
+      `incremental-engine: Ingestion '${id}': MARK ${
+        entities ? entities.length : 0
+      } entities, cursor: ${
+        cursor ? JSON.stringify(cursor) : 'none'
+      }, done: ${done}`,
+    );
+    const markId = v4();
+
+    await this.manager.createMark({
+      record: {
+        id: markId,
+        ingestion_id: id,
+        cursor,
+        sequence,
+      },
+    });
+
+    if (entities && entities.length > 0) {
+      await this.manager.createMarkEntities(markId, entities);
+    }
+
+    const added =
+      entities?.map(deferred => ({
+        ...deferred,
+        entity: {
+          ...deferred.entity,
+          metadata: {
+            ...deferred.entity.metadata,
+            annotations: {
+              ...deferred.entity.metadata.annotations,
+              [INCREMENTAL_ENTITY_PROVIDER_ANNOTATION]:
+                this.options.provider.getProviderName(),
+            },
+          },
+        },
+      })) ?? [];
+
+    const removed: DeferredEntity[] = [];
+
+    let doComputeRemoved = false;
+
+    if (!done) {
```

Comment:
I can't figure out this one I think 🤔 You want to issue the removals, only when it IS done, right? And you would probably want to either

1. Issue first just all the added (whenever there's anything at all in the `entities` array), and then issue all the removed separately, or
2. Issue additions and removals in one go, but then you would have to pass in the added array into `computeRemoved` so that it can exclude those from the list of things to remove

I think the first alternative would be the cleanest, and I don't think it would be significantly slower either.

Reply:
Yep, I got a bit turned around on this one. Fixed.